src/Native/Unix/System.Native/pal_signal.c

```diff
@@ -0,0 +1,287 @@
+    assert(arg != NULL);
+    int pipeFd = *(int*)arg;
+    free(arg);
+    assert(pipeFd >= 0);
+
+    // Continually read a signal code from the signal pipe and process it,
+    // until the pipe is closed.
+    while (true)
+    {
+        // Read the next signal, trying again if we were interrupted
+        uint8_t signalCode;
+        ssize_t bytesRead;
+        while ((bytesRead = read(pipeFd, &signalCode, 1)) < 0 && errno == EINTR);
+
+        if (bytesRead <= 0)
+        {
+            // Write end of pipe was closed or another error occurred.
+            // Regardless, no more data is available, so we close the read
+            // end of the pipe and exit.
+            close(pipeFd);
+            return NULL;
+        }
+
+        if (signalCode == SIGQUIT || signalCode == SIGINT)
+        {
+            // We're now handling SIGQUIT and SIGINT. Invoke the callback, if we have one.
+            CtrlCallback callback = g_ctrlCallback;
+            int rv = callback != NULL ? callback(signalCode == SIGQUIT ? Break : Interrupt) : 0;
+            if (rv == 0) // callback removed or was invoked and didn't handle the signal
+            {
+                // In general, we now want to remove our handler and reissue the signal to
+                // be picked up by the previously registered handler.  In the most common case,
+                // this will be the default handler, causing the process to be torn down.
+                // It could also be a custom handler registered by other code before us.
+                UninitializeConsole();
+                sigaction(signalCode, OrigActionFor(signalCode), NULL);
+                kill(getpid(), signalCode);
+            }
+        }
+        else if (signalCode == SIGCHLD)
+        {
+            SigChldCallback callback = g_sigChldCallback;
+            if (callback != NULL)
+            {
+                callback();
+            }
+            else
+            {
+                SystemNative_ResumeSigChld();
+            }
+        }
+        else
+        {
+            assert_msg(false, "invalid signalCode", (int)signalCode);
+        }
+    }
+}
+
+static void CloseSignalHandlingPipe()
+{
+    assert(g_signalPipe[0] >= 0);
+    assert(g_signalPipe[1] >= 0);
+    close(g_signalPipe[0]);
+    close(g_signalPipe[1]);
+    g_signalPipe[0] = -1;
+    g_signalPipe[1] = -1;
+}
+
+void SystemNative_RegisterForCtrl(CtrlCallback callback)
+{
+    assert(callback != NULL);
+    assert(g_ctrlCallback == NULL);
+    g_ctrlCallback = callback;
+}
+
+void SystemNative_UnregisterForCtrl()
+{
+    assert(g_ctrlCallback != NULL);
+    g_ctrlCallback = NULL;
+}
+
+void SystemNative_RegisterForSigChld(SigChldCallback callback)
+{
+    assert(callback != NULL);
+    assert(g_sigChldCallback == NULL);
+    g_sigChldCallback = callback;
+}
+
+static bool InstallSignalHandler(int sig, bool overwriteIgnored)
+{
+    int rv;
+    struct sigaction* orig = OrigActionFor(sig);
+
+    if (!overwriteIgnored)
```

Comment:
Nit: this name is a bit confusing to me... maybe "allowOverwritingIgnored", or something like that?